workspace/codespaces/codespaces.py

```diff
@@ -0,0 +1,88 @@
+import collections
+import datetime
+import json
+import os
+
+import requests
+
+from workspace.utils import blocks
+
+
+URL_PATTERN = "https://api.github.com/orgs/{org}/codespaces"
+TOKEN = os.environ["CODESPACES_GITHUB_API_TOKEN"]
+HEADERS = {
+    "Accept": "application/vnd.github+json",
+    "X-GitHub-Api-Version": "2022-11-28",
+}
+
+
+Codespace = collections.namedtuple(
+    "Codespace", ["age_in_days", "owner", "name", "has_unpushed_changes"]
+)
+
+
+def fetch(url, key):
+    """
+    Recursively fetch all records from the GitHub REST API endpoint given by `url`.
+    `key` is the property under which the records are stored.
+    """
+
+    def set_bearer_auth(request):
+        request.headers["Authorization"] = f"Bearer {TOKEN}"
+        return request
+
+    def fetch_one_page(page_url):
+        response = requests.get(page_url, auth=set_bearer_auth, headers=HEADERS)
+        response.raise_for_status()
+        yield from response.json().get(key)
+        if "next" in response.links:
+            yield from fetch_one_page(response.links["next"]["url"])
+
+    yield from fetch_one_page(url)
+
+
+def get_codespaces_for_org(org):
+    now = datetime.datetime.now(datetime.timezone.utc)
+    records = fetch(URL_PATTERN.format(org=org), "codespaces")
+    for record in records:
+        last_used_at = datetime.datetime.fromisoformat(record["last_used_at"])
+        age_in_days = (now - last_used_at).days
+        owner = record["owner"]["login"]
+        name = record["name"]
+        has_unpushed_changes = record["git_status"]["has_unpushed_changes"] == "true"
+        yield Codespace(age_in_days, owner, name, has_unpushed_changes)
+
+
+def is_at_risk(codespace, threshold_in_days):
```

Comment:
I think we should extend this to consider a Codespace at risk if it has uncommitted or unpushed changes as well. I expect quite a few Codespaces are created and then never used. However, I'd be happy to wait and see if that's the case.

Reply:
I think you mean the definition of "At risk" would be "Last used 20 or more days ago OR has unpushed changes OR has uncommitted changes". Is that correct?

I'm going to rename `Codespace.age_in_days` to `Codespace.last_used_days_ago` because "age" is confusing.

Reply:
If there are no uncommitted or unpushed changes are we actually concerned that a codespace is close to being deleted? What action would we take about those codespaces? I'm not sure we want to be chasing every researcher who didn't delete a throwaway codespace with no unsaved changes.

If we're not concerned about "normal" stale codespace deletion without unsaved changes, I suggest instead we define "at risk" to mean `over_threshold AND (has_unpushed OR has_uncommitted)`.

Reply:
That clarity could even extend to the function name

Reply:
Thanks for your comments. Let's adopt the definition of "at risk" that Mike suggested:

```
is_dormant >= 20 and (has_unpushed or has_uncommitted)
```